File: src/main/java/com/indiepost/repository/jpa/RoleRepositoryJpa.java
package com.indiepost.repository.jpa;

import com.indiepost.enums.Types.UserRole;
import com.indiepost.model.QRole;
import com.indiepost.model.Role;
import com.indiepost.repository.RoleRepository;
import com.querydsl.jpa.impl.JPAQueryFactory;
import org.springframework.stereotype.Repository;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;

/**
 * Created by jake on 8/4/16.
 */
@Repository
public class RoleRepositoryJpa implements RoleRepository {

    static final QRole r = QRole.role;

    @PersistenceContext
    private EntityManager entityManager;

    @Override
    public void save(UserRole role) {
        entityManager.persist(role);
    }

    @Override
    public void update(UserRole role) {
        entityManager.persist(role);
    }

    @Override
    public void delete(UserRole role) {
        entityManager.remove(role);
    }

    @Override
    public Role findById(Long id) {
        return getQueryFactory()
                .selectFrom(r)
                .where(r.id.eq(id))
                .fetchOne();
    }

    @Override
    public Role findByUserRole(UserRole role) {
        return findByUserRoleString(role.toString());
    }

    @Override
    public Role findByUserRoleString(String role) {
        return getQueryFactory()
                .selectFrom(r)
                .where(r.name.eq(role))
                .fetchOne();
    }

    private JPAQueryFactory getQueryFactory() {
        return new JPAQueryFactory(entityManager);
    }
}

File: src/main/kotlin/com/indiepost/repository/jpa/RoleRepositoryJpa.kt
package com.indiepost.repository.jpa

import com.indiepost.enums.Types.UserRole
import com.indiepost.model.QRole
import com.indiepost.model.Role
import com.indiepost.repository.RoleRepository
import com.querydsl.jpa.impl.JPAQueryFactory
import org.springframework.stereotype.Repository

import javax.persistence.EntityManager
import javax.persistence.PersistenceContext

/**
 * Created by jake on 8/4/16.
 */
@Repository
class RoleRepositoryJpa : RoleRepository {

    @PersistenceContext
    private lateinit var entityManager: EntityManager

    private val r = QRole.role

    private val queryFactory: JPAQueryFactory
        get() = JPAQueryFactory(entityManager)

    override fun save(role: UserRole) {
        entityManager.persist(role)
    }

    override fun update(role: UserRole) {
        entityManager.persist(role)
    }

    override fun delete(role: UserRole) {
        entityManager.remove(role)
    }

    override fun findById(id: Long?): Role? {
        return queryFactory
                .selectFrom(r)
                .where(r.id.eq(id))
                .fetchOne()
    }

    override fun findByUserRole(role: UserRole): Role? {
        return findByUserRoleString(role.toString())
    }

    override fun findByUserRoleString(role: String): Role? {
        val userRole = UserRole.valueOf(role)
        return queryFactory
                .selectFrom(r)
                .where(r.roleType.eq(userRole))
                .fetchOne()
    }
}
